bpo-33671 / shutil.copyfile: use memoryview() with dynamic size on Windows

giampaolo opened the pull request:
This is a follow up of https://github.com/python/cpython/pull/7160.
Changes:
- use `memoryview()` with size == file size, see https://github.com/python/cpython/pull/7160#discussion_r195405230
- release intermediate (sliced) `memoryview` immediately
- replace "OSX" occurrences with "macOS"
- add some unittests for copyfileobj()
- update doc

<!-- issue-number: bpo-33671 -->
https://bugs.python.org/issue33671
<!-- /issue-number -->

giampaolo committed:
bpo-33671: release sub-memoryview resources immediately

giampaolo committed:
rename OSX to macOS

giampaolo committed:
add unit tests for copyfileobj

giampaolo committed:
Remove _copybinfileobj() implementation

...as it's considerably slower for smaller files; e.g. with a 8MiB file:

--- original

~/svn/cpython {33671-release-memoryview}$ ./python  -m timeit -s "import
shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')"
"shutil.copyfileobj(f1, f2)"
200000 loops, best of 5: 1.64 usec per loop

--- _copybinfileobj()

~/svn/cpython {33671-release-memoryview}$ ./python  -m timeit -s "import
shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')"
"shutil.copyfileobj(f1, f2)"
100000 loops, best of 5: 3.3 usec per loop

giampaolo committed:
update doc

giampaolo committed:
update doc

giampaolo committed:
update doc

giampaolo committed:
update doc

giampaolo committed:
update doc

giampaolo committed:
Reintroduce memoryview() variant of copyfileobj()

...and use it from copyfile() only if WINDOWS and file size >= 128 MiB

giampaolo committed:
speed up test case by mocking file size instead of creating an  actual 128 ;iB file

giampaolo committed:
update doc

giampaolo committed:
update comments

giampaolo committed:
rename var

vstinner commented on Lib/shutil.py:
```diff
@@ -234,26 +229,36 @@ def copyfile(src, dst, *, follow_symlinks=True):
-
-            if _HAS_FCOPYFILE:
+            # macOS
+            elif _HAS_FCOPYFILE:
```
Both options are always exclusive? macOS might have sendfile() in the future, no?

giampaolo replied:
This is now outdated: I updated code to always use fcopyfile on OSX and leave sendfile() alone. This way we avoid invoking sendfile() on OSX which would fail anyway.

vstinner commented on Lib/shutil.py:
```diff
@@ -234,26 +229,36 @@ def copyfile(src, dst, *, follow_symlinks=True):
+            # considerable speedup by using a readinto()/memoryview()
+            # variant of copyfileobj(), see:
+            # https://github.com/python/cpython/pull/7160#discussion_r195162475
+            elif _WINDOWS and file_size >= 128 * 1024 * 1024:
```
Would you mind to add a top-level private constant rather than an hardcoded value? (Would it make sense to make it somehow public? Or add an optional parameter...?)

giampaolo replied:
This is outdated, I modified code to always use the readinto() variant regardless of file size, see: https://github.com/python/cpython/pull/7160#discussion_r195458353

giampaolo committed:
Use dynamic memoryview() size based on file size

As per https://github.com/python/cpython/pull/7160#discussion_r195405230
it appears using memoryview() on Windows is faster also for small files
if memoryview() length is equal to file size.

giampaolo committed:
do not attempt to use sendfile() on OSX
